keps/sig-api-machinery/20191210-consistent-reads-from-cache.md

```diff
@@ -0,0 +1,404 @@
+this by configuring reflectors that would benefit most from this feature, like
+the pod list requests from kubelet pod, to not paginate.
+
+#### Option: Serve 1st page of paginated requests from the watch cache
+
+Only serve the 1st page of paginated requests from the watch cache. The watch
+cache would need to construct the appropriate continuation token such that the
+subsequent pages can be served from etcd.
+
+An even more conservative approach would be to only serve paginated requests
+that fit within a single page from the watch cache, in which cache the watch
+cache doesn't need to construct continuation tokens at all.
+
+In practice, this options might be sufficient to get the bulk of the scalability
+benefits of serving consistent reads from cache.
+
+#### Rejected Option: Return unpaginated responses to paginated list requests
+
+Both for backward compatibility with older versions of Kubernetes that do not support
+pagination, and for compatibility with api-servers that have the watch cache enabled,
+clients must be able to tolerate unpaginated responses for paginated LIST requests.
+
+The kube-apiserver already switches between serving paginated responses and
+serving unpaginated responses depending on if the watch cache is enabled for the
+types requested. 
+
+User cases that have disabled the watch cache will still receive paginated responses.
+
+Use cases where paginated was being used to reduce load on etcd, but are able to
+tolerate the data volume of the list being returned in a single response, should
+not impacted.
+
+The most likely problem with this approach is that, because LIST with
+resourceVersion="" is the only way to paginate data from the api-server when the
+watch-cache is enabled (which is the default), clients that need pagination
+(e.g. to avoid receiving too much data in a single request) will be relying on
+LIST resourceVersion="".
+
+We are not planning to pursue this option.
+
+#### Rejected Option: Enable pagination in the watch cache
+
+The problem is that the watch cache ("isn't able to perform
+continuations")[https://github.com/kubernetes/kubernetes/blob/789dc873f6816cc2b9b39e77a9b94f478d3a3134/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L595].
+The watch cache is designed to only serve LIST requests at [the latest resource version is has
+available](https://github.com/kubernetes/kubernetes/blob/789dc873f6816cc2b9b39e77a9b94f478d3a3134/staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go#L115).
+
+To supporting watch cache pagination:
+- The watch cache would to keep a comparable resource version history to the
+  default etcd compaction history of 5 minutes.
+- The watch cache would need to be resturctured so that is can serve LIST for
+  the resource versions it has returned continuation tokens to clients for.
+
+Both of these are major changes, and would require scalability validation. We
+are not planning to pursue this option.
```

Comment:
Actually, I was thinking about that and maybe we can do some poor-man support for pagination.
Consider the following:
- watch cache is getting LIST request with pagination
- we list everything from the internal cache and have pointers to those objects
- we return first LIMIT of those, and in the internal map store the remaining ones indexed by "continuation token" that we just generated
- we GC the items from this map after N seconds from insertion
- if continuation is set in the request, we lookup that map and return next LIMIT items, if the item doesn't exist in the map we (either fallback to etcd or return an error - probably the former)

These kind of changes should be fairly simple and maybe those actually be enough?

@lavalamp @liggitt @jpbetz - for thoughts

Reply:
Determining good value for N (seconds after which items are garbage collected) may be tricky.
Another measure is to bound the memory consumed by this map.

Reply:
Something like this could work. We'd need to think through out the memory utilization/bounds. I'll add a note and we can expand on it.